File: CountPrimes.py
#Prime numbers are natural numbers greater than 1 that 
# have no divisors other than 1 and themselves
# Examples of prime numbers: 2, 3, 5, 7, 11, 13, 17, 19, 23...
import math


class Solution:
    def countPrimes(self, n: int, i=None) -> int:
        # create a list of size n and has a default value of true
        if (n <= 2):
            return 0
        isPrime = [True] * (n)
        isPrime[0] = isPrime[1] = False
        for i in range(2, int(math.sqrt(n)) + 1):
            if isPrime[i]:
                for j in range(i * i, n, i):
                    isPrime[j] = False

        count = 0
        for i in range(2, n):
            if (isPrime[i]):
                count = count + 1
        return count

File: CountPrimes.cpp
#include <iostream>
#include <vector>
#include <math.h>
using namespace std;

class Solution
{
  //* Implememting the most effeccient solution for calculating the primes count
  // using the Sieve Sieve of Eratosthenes
public:
  int countPrimes(int n)
  {
    int count = 0;
    vector<bool> isPrime(n, true);
    for (int i = 2; i <= sqrt(n); ++i)
    {
      // coreLogic: if the number i is a prime, then surely its multiples arenot primes
      if (isPrime[i])
      {
        for (int j = i * i; j < n; j += i)
        {
          isPrime[j] = false;
        }
      }
    }
    for (int i = 2; i < n; ++i)
    {
      if (isPrime[i])
      {
        count += 1;
      }
    }
    return count;
  }
};

int main()
{
  Solution sol;
  cout << sol.countPrimes(10);
  return 0;
}